File: js/tilings/pentagon.js
/**
 * Provides various pentagonal tilings.
 */

export function generateCairoPentagonBoard(options) {
    return generatePentagonBoardByEdges(options, "cairo");
}

export function generatePrismaticPentagonBoard(options) {
    return generatePentagonBoardByEdges(options, "prismatic");
}

/**
 * Cairo pentagonal tiling is a dual of the snub square tiling.
 * We implement it using a square grid where each cell is divided into 2 pentagons,
 * alternating orientation in a checkerboard pattern.
 * Midpoints are added to all edges to ensure perfect tiling and neighbor detection.
 */
export function generatePentagonBoardByEdges(options, type) {
    const { cols, rows, tileSize: s, colorCount, rng } = options;
    const rawTiles = [];
    const m = s / 2;

    if (type === "cairo") {
        const d = 0.2 * s; // Depth of the "V" division
        for (let r = 0; r < rows; r++) {
            for (let c = 0; c < cols; c++) {
                const x = c * s;
                const y = r * s;

                if ((r + c) % 2 === 0) {
                    // Horizontal division (Top and Bottom)
                    rawTiles.push({
                        points: [
                            [x, y], [x + m, y], [x + s, y],      // Top edge
                            [x + s, y + m],                     // Right midpoint
                            [x + m, y + m - d],                // Peak
                            [x, y + m]                          // Left midpoint
                        ]
                    });
                    rawTiles.push({
                        points: [
                            [x, y + m], [x + m, y + m - d], [x + s, y + m], // Midline
                            [x + s, y + s], [x + m, y + s], [x, y + s]      // Bottom edge
                        ]
                    });
                } else {
                    // Vertical division (Left and Right)
                    rawTiles.push({
                        points: [
                            [x, y], [x + m, y],                 // Top midpoint
                            [x + m - d, y + m],                // Peak
                            [x + m, y + s], [x, y + s],         // Bottom midpoint
                            [x, y + m], [x, y]                  // Left edge
                        ]
                    });
                    rawTiles.push({
                        points: [
                            [x + m, y], [x + s, y], [x + s, y + m], [x + s, y + s], // Right edge
                            [x + m, y + s],                     // Bottom midpoint
                            [x + m - d, y + m]                 // Peak
                        ]
                    });
                }
            }
        }
    } else {
        // Prismatic (House) tiling: Interlocking
        const h = s * 0.7; // height of the square part
        const p = s * 0.3; // height of the peak

        for (let r = 0; r < rows; r++) {
            const isOdd = r % 2 === 1;
            const yBase = r * h;
            for (let c = 0; c < cols; c++) {
                const x = c * s + (isOdd ? s/2 : 0);

                if (!isOdd) {
                    // Downward house
                    rawTiles.push({
                        points: [
                            [x, yBase], [x + m, yBase], [x + s, yBase], // Top
                            [x + s, yBase + m], [x + s, yBase + h],     // Right
                            [x + m, yBase + h + p],                     // Peak
                            [x, yBase + h], [x, yBase + m]              // Left
                        ]
                    });
                } else {
                    // Upward house
                    rawTiles.push({
                        points: [
                            [x + m, yBase - p],                         // Peak
                            [x + s, yBase], [x + s, yBase + m], [x + s, yBase + h], // Right
                            [x + m, yBase + h], [x, yBase + h],         // Bottom
                            [x, yBase + m], [x, yBase]                  // Left
                        ]
                    });
                }
            }
        }
    }

    // Determine bounds to offset
    let minX = Infinity, minY = Infinity;
    let maxX = -Infinity, maxY = -Infinity;
    for (const t of rawTiles) {
        for (const p of t.points) {
            minX = Math.min(minX, p[0]);
            minY = Math.min(minY, p[1]);
            maxX = Math.max(maxX, p[0]);
            maxY = Math.max(maxY, p[1]);
        }
    }

    const tiles = rawTiles.map((t, i) => ({
        id: i,
        colorId: Math.floor(rng() * colorCount),
        ownerId: null,
        points: t.points.map(p => [p[0] - minX, p[1] - minY]),
        neighbors: []
    }));

    // Shared edge neighbor detection
    const edgeMap = new Map();
    const quantize = (v) => Math.round(v * 100) / 100;

    tiles.forEach(tile => {
        for (let i = 0; i < tile.points.length; i++) {
            const p1 = tile.points[i];
            const p2 = tile.points[(i + 1) % tile.points.length];
            const x1 = quantize(p1[0]), y1 = quantize(p1[1]);
            const x2 = quantize(p2[0]), y2 = quantize(p2[1]);

            // Sort points to make the edge key stable
            const key = x1 < x2 || (x1 === x2 && y1 < y2)
                ? `${x1},${y1}_${x2},${y2}`
                : `${x2},${y2}_${x1},${y1}`;

            if (!edgeMap.has(key)) edgeMap.set(key, []);
            edgeMap.get(key).push(tile.id);
        }
    });

    edgeMap.forEach(tileIds => {
        if (tileIds.length >= 2) {
            for (let i = 0; i < tileIds.length; i++) {
                for (let j = i + 1; j < tileIds.length; j++) {
                    const id1 = tileIds[i];
                    const id2 = tileIds[j];
                    if (!tiles[id1].neighbors.includes(id2)) tiles[id1].neighbors.push(id2);
                    if (!tiles[id2].neighbors.includes(id1)) tiles[id2].neighbors.push(id1);
                }
            }
        }
    });

    return {
        version: 1,
        generator: `pentagon-${type}`,
        width: maxX - minX,
        height: maxY - minY,
        tiles,
        startTileIds: [0, tiles.length - 1]
    };
}

export function generatePentagonBoard(options) {
    return generateCairoPentagonBoard(options);
}

File: go-flood/pkg/tilings/pentagon.go
package tilings

import (
	"fmt"
	"go-flood/pkg/core"
	"math"
)

func GenerateCairoPentagonBoard(options Options) core.Board {
	size := options.TileSize
	hSize := size / 2
	offset := size * 0.2
	cols, rows := options.Cols, options.Rows

	getP1 := func(gx, gy int) core.Point {
		x := float64(gx) * size
		y := float64(gy) * size
		if (gx+gy)%2 == 0 {
			return core.Point{x + hSize, y + offset}
		}
		return core.Point{x + offset, y + hSize}
	}
	getP2 := func(gx, gy int) core.Point {
		x := float64(gx) * size
		y := float64(gy) * size
		if (gx+gy)%2 == 0 {
			return core.Point{x + hSize, y + size - offset}
		}
		return core.Point{x + size - offset, y + hSize}
	}
	getCorner := func(gx, gy int) core.Point {
		return core.Point{float64(gx) * size, float64(gy) * size}
	}

	type cairoTemp struct {
		core.Tile
		gx, gy, subId int
	}
	tempTiles := []cairoTemp{}

	addTile := func(points []core.Point, gx, gy, subId int) {
		id := (gy*cols+gx)*2 + subId
		colorId := int(options.RNG() * float64(options.ColorCount))
		tempTiles = append(tempTiles, cairoTemp{
			Tile: core.Tile{
				ID:        id,
				ColorID:   colorId,
				OwnerID:   nil,
				Neighbors: []int{},
				Points:    points,
			},
			gx: gx, gy: gy, subId: subId,
		})
	}

	for gy := 0; gy < rows; gy++ {
		for gx := 0; gx < cols; gx++ {
			isV := (gx+gy)%2 == 0
			p1 := getP1(gx, gy)
			p2 := getP2(gx, gy)

			if isV {
				// Left
				addTile([]core.Point{
					p1,
					getCorner(gx, gy),
					getP2(gx-1, gy),
					getCorner(gx, gy+1),
					p2,
				}, gx, gy, 0)
				// Right
				addTile([]core.Point{
					p1,
					getCorner(gx+1, gy),
					getP1(gx+1, gy),
					getCorner(gx+1, gy+1),
					p2,
				}, gx, gy, 1)
			} else {
				// Top
				addTile([]core.Point{
					p1,
					getCorner(gx, gy),
					getP2(gx, gy-1),
					getCorner(gx+1, gy),
					p2,
				}, gx, gy, 0)
				// Bottom
				addTile([]core.Point{
					p1,
					getCorner(gx, gy+1),
					getP1(gx, gy+1),
					getCorner(gx+1, gy+1),
					p2,
				}, gx, gy, 1)
			}
		}
	}

	tileMap := make(map[string]int)
	for i, t := range tempTiles {
		tileMap[fmt.Sprintf("%d,%d,%d", t.gx, t.gy, t.subId)] = i
	}

	for i := range tempTiles {
		t := &tempTiles[i]
		gx, gy, subId := t.gx, t.gy, t.subId
		isV := (gx+gy)%2 == 0
		var potential [][3]int

		if isV {
			if subId == 0 { // Left
				potential = [][3]int{
					{gx, gy, 1},
					{gx - 1, gy, 0},
					{gx - 1, gy, 1},
					{gx, gy - 1, 1},
					{gx, gy + 1, 0},
				}
			} else { // Right
				potential = [][3]int{
					{gx, gy, 0},
					{gx + 1, gy, 0},
					{gx + 1, gy, 1},
					{gx, gy - 1, 1},
					{gx, gy + 1, 0},
				}
			}
		} else {
			if subId == 0 { // Top
				potential = [][3]int{
					{gx, gy, 1},
					{gx, gy - 1, 0},
					{gx, gy - 1, 1},
					{gx - 1, gy, 1},
					{gx + 1, gy, 0},
				}
			} else { // Bottom
				potential = [][3]int{
					{gx, gy, 0},
					{gx, gy + 1, 0},
					{gx, gy + 1, 1},
					{gx - 1, gy, 1},
					{gx + 1, gy, 0},
				}
			}
		}

		for _, p := range potential {
			if idx, ok := tileMap[fmt.Sprintf("%d,%d,%d", p[0], p[1], p[2])]; ok {
				t.Neighbors = append(t.Neighbors, tempTiles[idx].ID)
			}
		}
	}

	minX, minY := math.MaxFloat64, math.MaxFloat64
	maxX, maxY := -math.MaxFloat64, -math.MaxFloat64

	for _, t := range tempTiles {
		for _, p := range t.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}

	finalTiles := make([]core.Tile, len(tempTiles))
	for i, t := range tempTiles {
		newPoints := make([]core.Point, len(t.Points))
		for j, p := range t.Points {
			newPoints[j] = core.Point{p[0] - minX, p[1] - minY}
		}
		t.Points = newPoints
		finalTiles[i] = t.Tile
	}

	return core.Board{
		Version:      1,
		Generator:    "pentagon-cairo",
		Width:        maxX - minX,
		Height:       maxY - minY,
		Tiles:        finalTiles,
		StartTileIds: []int{0, len(finalTiles) - 1},
	}
}
